compiler/rustc_errors/src/lib.rs

```diff
@@ -1295,20 +1299,15 @@ impl DiagCtxtInner {
                 return None;
             }
             Warning if !self.flags.can_emit_warnings => {
-                if diagnostic.has_future_breakage() {
-                    (*TRACK_DIAGNOSTIC)(diagnostic, &mut |_| {});
-                }
                 return None;
             }
             Allow | Expect(_) => {
-                (*TRACK_DIAGNOSTIC)(diagnostic, &mut |_| {});
                 return None;
             }
             _ => {}
         }
 
-        let mut guaranteed = None;
-        (*TRACK_DIAGNOSTIC)(diagnostic, &mut |mut diagnostic| {
+        TRACK_DIAGNOSTIC(diagnostic, &mut |mut diagnostic| {
```

Comment:
Could you add a comment on this function explaining that we only need to track diagnostics that are actually emitted?